rfc-116-store-attachment-data-in-content-items.md

```diff
@@ -0,0 +1,110 @@
+# Store attachment data in content items
+
+## Summary
+
+Add a new field to the details hash of content items, `attachments`,
+which has metadata about the page's attachments (if any).
+
+
+## Problem
+
+Pages on GOV.UK can have attachments, which come in a few different
+types.  Whitehall, which probably has the richest model of
+attachments, has:
+
+- External attachments (links to other websites)
+- HTML attachments
+- File attachments (which can have previews)
+
+Attachments are referenced in the content of the page, and make up
+part of the govspeak (or HTML) the publishing app sends to the
+publishing-api.
+
+Attachments are *not* referenced anywhere else in the content item.
+To get the details of an attachment, you have to parse the body of the
+page.  This restricts what we can do with attachments.  For example,
+we cannot generate comprehensive [schema.org][] metadata for
+attachments: [see this comment][].
+
+Additionally, users of the content API cannot make use of attachments
+without parsing the page body.  This inhibits creative use of our
+content.
+
+
+## Proposal
+
+In our content schemas we already have an [`asset_link` type][], which
+[Specialist Publisher uses for storing its attachment data][].  This
+solves all of our problems: [Publishing API can render such
+attachments with govspeak][], and content items can expose this data
+to enable programmatic use.
+
+However, the metadata Govspeak expects and the `asset_link` type are
+*informally* the same.  And furthermore, Whitehall has much richer
+attachment metadata than either `asset_link` or Govspeak allow.  So
+this proposal has three parts:
+
+1. Add a new required field called `attachments` (of type
+   `asset_link_list`) to the `details` of formats which can have
+   attachments.
+
+2. Document the relationship between the `asset_link` type in our
+   content schemas and the data Govspeak attachments use.  For
+   example, all fields required in one must be required in the other.
+   Any pairs of fields which do the same job in both should have the
+   same name.  Implement any changes arising from this.
+
+3. Add new optional fields to the `asset_link` type:
+
+   - `file_size` (used by govspeak)
+   - `preview_url` (used by whitehall)
+
```

Comment:
A piece of data we're currently missing about attachments is language. It would be very helpful to be able to mark links to foreign language content appropriately for assistive tech, and also to be able to sort attachments so that if there are 2 language versions of a page we show only the correctly associated files alongside it. It looks like this might be a place this could be implemented.

Reply:
This is a good idea, however we unfortunately we don't know whether the language of an attachment has deviated from the content of the document as a user never inputs this information.

I will raise this though in the Publishing Workflow team as we design the process for publication attachment management to see if it's something that can be improved.

Reply:
Whitehall does have a `locale` field for attachments.  I'm not sure how valuable it is, due to Kevin's comment.  I've included it (along with all the other Whitehall metadata) in the updated draft.